### PR TITLE
haskellPackages.hail: add patches to relax cabal dependencies

### PR DESCRIPTION
**Pull request by astro**
###### Motivation for this change
Didn't build.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @ehmry

### CODE REVIEW

**Review by ghost — APPROVED** [review]

**Review by cdepillabout — CHANGES_REQUESTED** [review]
@astro In order to get this merged, could you do the following things:

1. For each patch, add a link to the PR it comes from and a short description of why it is needed (what it does).
2. Ask upstream to make a new release of `hail` with these patches pulled in.  Since these patches appear to only be dependency version fixes, it could just be a Hackage revision instead of a full release.  If you added a comment to this PR explaining this, and saying which version of `hail` you expect to be working without applying these patches, it would be helpful for the nixpkgs maintainers to know when we can remove these patches.
3. Remove this line: https://github.com/NixOS/nixpkgs/blob/2e52f98ddbe9ddd91f65f8df471d84e1e4077bdf/pkgs/development/haskell-modules/configuration-hackage2nix.yaml#L5356.  You can find an explanation of this here: https://discourse.nixos.org/t/video-tutorial-how-to-fix-broken-haskell-packages-in-nix/3968.
4. Change the base branch for this PR to `haskell-updates` instead of `master`.  Then actually rebase the commit in this PR onto the `haskell-updates` branch.

**Comment by tfc** [comment]
The first patch is already merged to master, and the second patch is a PR that just needs to be merged. how about contacting @shlevy about this to have both patches in and then simply bumping `src` up to the latest commit on the master branch?

I would profit from this too, as we use hail in our intranet.

**Comment by astro** [comment]
@cdepillabout Thank you for your writeup. I hope to have addressed all concerns.

**Comment by astro** [comment]
> The first patch is already merged to master, and the second patch is a PR that just needs to be merged. how about contacting @shlevy about this to have both patches in and then simply bumping src up to the latest commit on the master branch?

Ok, I have asked in the 2nd PR.

We still would love to have unbroken Hail in 19.09 rather soon than waiting for a single maintainer.

File: pkgs/development/haskell-modules/configuration-common.nix
Comment (cdepillabout) [inline pkgs/development/haskell-modules/configuration-common.nix]: Comments here look good.

File: pkgs/development/haskell-modules/configuration-hackage2nix.yaml
Comment (cdepillabout) [inline pkgs/development/haskell-modules/configuration-hackage2nix.yaml]: You don't actually need to regenerate the `default-package-overrides` part of this file.

In general, @peti will do this once a week a or so.

File: pkgs/development/haskell-modules/configuration-hackage2nix.yaml
Comment (cdepillabout) [inline pkgs/development/haskell-modules/configuration-hackage2nix.yaml]: All you need to do is remove this line from the `broken-packages` section.

File: pkgs/development/haskell-modules/hackage-packages.nix
Comment (cdepillabout) [inline pkgs/development/haskell-modules/hackage-packages.nix]: This file shouldn't be modified either, since @peti does it once a week or so.

File: pkgs/development/haskell-modules/configuration-hackage2nix.yaml
Comment (astro) [reply pkgs/development/haskell-modules/configuration-hackage2nix.yaml]: Oops, the rest slipped in due to a last test build.

Comment (astro) [reply pkgs/development/haskell-modules/configuration-hackage2nix.yaml]: Actually, no. I need to update the PR's target branch.

**Comment by cdepillabout** [comment]
@astro Thanks for updating this.

However, you'll need to do the following things before this can be merged in:

1. Change the base branch for this PR to `haskell-updates` instead of `master`. Then actually rebase the commit in this PR onto the `haskell-updates` branch.
2. Make sure not to modify the `pkgs/development/haskell-modules/hackage-packages.nix` file, since this is done manually once-a-week or so by @peti.
3. Make sure not to regenerate the `default-package-overrides` section of the `pkgs/development/haskell-modules/configuration-hackage2nix.yaml` file, since this is also done by @peti.

> We still would love to have unbroken Hail in 19.09 rather soon than waiting for a single maintainer.

As soon as the above are fixed, I think this should be fine to be merged in.  Although when a new release of `hail` is made, we'd be very grateful if you could send another PR removing the patches added in this PR.  (I imagine the patches will fail to apply, `hail` will fail to compile, and it might be marked `broken` again.)

**Comment by cdepillabout** [comment]
@astro Okay, it looks like you fixed everything :-)

I'll merge after CI passes.

**Comment by cdepillabout** [comment]
Okay, CI has passed, so merging in.

@astro Thanks again.  This should be available when the haskell package set has been regenerated (usually in a week or so).

**Comment by shlevy** [comment]
Will upload a fixed version of hail to hackage, sorry!

**Pull request merged**
